storage/bootstrap/bootstrapper/commitlog/source_prop_test.go

```diff
@@ -0,0 +1,235 @@
+			bootstrapOpts := testOptions().SetCommitLogOptions(commitLogOpts)
+
+			// Instantiate commitlog
+			log, err := commitlog.NewCommitLog(commitLogOpts)
+			if err != nil {
+				return false, err
+			}
+			err = log.Open()
+			if err != nil {
+				return false, err
+			}
+
+			// Write all the datapoints to the commitlog
+			for _, write := range input.writes {
+				log.Write(context.NewContext(), write.series, write.datapoint, write.unit, write.annotation)
+			}
+			err = log.Close()
+			if err != nil {
+				return false, err
+			}
+
+			// Instantiate a commitlog source
+			source := newCommitLogSource(bootstrapOpts)
+
+			// Determine time range to bootstrap
+			md := testNsMetadata(t)
+			blockSize := md.Options().RetentionOptions().BlockSize()
+			start := input.currentTime.Truncate(blockSize)
+			end := input.currentTime.Add(blockSize)
+			ranges := xtime.NewRanges()
+			ranges = ranges.AddRange(xtime.Range{
+				Start: start,
+				End:   end,
+			})
+
+			// Determine which shards we need to bootstrap (based on the randomly
+			// generated data)
+			allShards := map[uint32]bool{}
+			for _, write := range input.writes {
+				allShards[write.series.Shard] = true
+			}
+
+			// Assign the previously-determined bootstrap range to each known shard
+			shardTimeRanges := result.ShardTimeRanges{}
+			for shard := range allShards {
+				shardTimeRanges[shard] = ranges
+			}
+
+			// Perform the bootstrap
+			result, err := source.Read(md, shardTimeRanges, testDefaultRunOpts)
+			if err != nil {
+				return false, err
+			}
+
+			// Create testValues for each datapoint for comparison
+			values := []testValue{}
+			for _, write := range input.writes {
+				values = append(values, testValue{write.series, write.datapoint.Timestamp, write.datapoint.Value, write.unit, write.annotation})
+			}
+
+			err = verifyShardResultsAreCorrect(values, result.ShardResults(), bootstrapOpts)
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		},
+		genPropTestInputs(testNamespaceID.String()),
+	))
+
+	props.TestingRun(t)
+}
+
+type propTestInput struct {
+	currentTime time.Time
+	writes      []generatedWrite
+}
+
+type generatedWrite struct {
+	series     commitlog.Series
+	datapoint  ts.Datapoint
+	unit       xtime.Unit
+	annotation ts.Annotation
+}
+
+func (w generatedWrite) String() string {
+	return fmt.Sprintf("ID = %v, Datapoint = %+v", w.series.ID.String(), w.datapoint)
+}
+
+func genPropTestInputs(ns string) gopter.Gen {
+	curriedGenPropTestInput := func(input interface{}) gopter.Gen {
+		inputs := input.([]interface{})
+		start := inputs[0].(time.Time)
+		numDatapoints := inputs[1].(int)
+		return genPropTestInput(start, numDatapoints, ns)
+	}
+	return gopter.CombineGens(
+		// Runs iterations of the test starting 1000 hours in the past/future
+		gen.TimeRange(time.Now().Add(-1000*time.Hour), 1000*time.Hour),
+		// Run iterations of the test with between 0 and 1000 datapoints
```

Comment:
maybe have a higher lower bound; [100,1000] instead of [0, 1000]?

Reply:
i actually thought zero datapoints was an interesting edge-case to include

Reply:
Cool, fair enough